### PR TITLE
fix: slither errors

### LINKED ISSUE #536: Fix slither CI errors - II
This workflow https://github.com/ubiquity/ubiquity-dollar/blob/6e891801df973a6613b43c899381cdd244ad3fdb/.github/workflows/slither-analysis.yml produces some errors

You can find the failed CI run here: https://github.com/ubiquity/ubiquity-dollar/actions/runs/4065041113/jobs/6999216155

What should be done:
- update contracts to fix slither errors and make CI action pass

### PR DESCRIPTION
**Pull request by ghost**
Resolves #536

this pull request is to resolve the slither errors which was coming from
yarn workspace @ubiquity/contracts run test:slither

### CODE REVIEW

**Comment by ghost** [comment]
@rndquu hm all the checks passed

**Comment by rndquu** [comment]
> @rndquu hm all the checks passed

@AnakinSkywalkeer could you reenable the [slither ci workflow](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/workflows/slither-analysis.yml) (i.e. add "pull_request" and "push" handlers) so that we could check how it works

**Comment by ghost** [comment]
> > @rndquu hm all the checks passed
> 
> @AnakinSkywalkeer could you reenable the [slither ci workflow](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/workflows/slither-analysis.yml) (i.e. add "pull_request" and "push" handlers) so that we could check how it works

sure

**Comment by ghost** [comment]
> > @rndquu hm all the checks passed
> 
> @AnakinSkywalkeer could you reenable the [slither ci workflow](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/workflows/slither-analysis.yml) (i.e. add "pull_request" and "push" handlers) so that we could check how it works

um ah i didn't get how to enable it hm 🤔

**Comment by ghost** [comment]
isn't enabling workflow requires admin perms

File: packages/contracts/scripts/deploy/dollar/solidityScripting/00_Constants.s.sol
Comment (0x4007) [inline packages/contracts/scripts/deploy/dollar/solidityScripting/00_Constants.s.sol:91]: This is nice and succinct but I presume that real money is on the line here and we should be especially careful to validate that these were all converted correctly.

File: packages/contracts/slither.config.json
Comment (0x4007) [inline packages/contracts/slither.config.json]: > state-variables-that-could-be-declared-constant

What is this doing? Why shouldn't state variables normally be declared as constants?

File: packages/contracts/src/dollar/facets/StakingFormulasFacet.sol
Comment (0x4007) [inline packages/contracts/src/dollar/facets/StakingFormulasFacet.sol]: Why was this and some of the other smart contracts downgraded?

File: packages/contracts/src/dollar/utils/CollectableDust.sol
Comment (0x4007) [inline packages/contracts/src/dollar/utils/CollectableDust.sol:1]: Isn't this an important contract? I'm pretty sure I've called collectable dust manually several times in the past using the admin wallet.

File: packages/contracts/slither.config.json
Comment (ghost) [reply packages/contracts/slither.config.json]: like i did make the most of the static variables constant but like there was these two on making them constant there was errors about "value should be known at compile time" so i had to use that to avoid it we can fix it if u want 
 uint256 deployerPrivateKey = vm.envUint("PRIVATE_KEY");
 address admin = vm.envAddress("PUBLIC_KEY");

File: packages/contracts/src/dollar/facets/StakingFormulasFacet.sol
Comment (ghost) [reply packages/contracts/src/dollar/facets/StakingFormulasFacet.sol]: > Why was this and some of the other smart contracts downgraded?

oh my bad there was like this errors about incorrect solidity version to check if i can downgrade and fix them i tried doing it but seems like i forget to make some of them to 0.8.19 i will fix that

File: packages/contracts/src/dollar/utils/CollectableDust.sol
Comment (ghost) [reply packages/contracts/src/dollar/utils/CollectableDust.sol:1]: > Isn't this an important contract? I'm pretty sure I've called collectable dust manually several times in the past using the admin wallet.

i was getting dead code error
CollectableDust (src/dollar/utils/CollectableDust.sol#11-58) does not implement functions:
- ICollectableDust.addProtocolToken(address) (src/dollar/interfaces/utils/ICollectableDust.sol#9)
- ICollectableDust.removeProtocolToken(address) (src/dollar/interfaces/utils/ICollectableDust.sol#11)
- ICollectableDust.sendDust(address,address,uint256) (src/dollar/interfaces/utils/ICollectableDust.sol#13)

i asked @rndquu they said that file is not in use so we can try removing it and see if it breaks anything it didn't so i went with that

File: packages/contracts/scripts/deploy/dollar/solidityScripting/00_Constants.s.sol
Comment (ghost) [reply packages/contracts/scripts/deploy/dollar/solidityScripting/00_Constants.s.sol:91]: > This is nice and succinct but I presume that real money is on the line here and we should be especially careful to validate that these were all converted correctly.

yes for sure

File: packages/contracts/slither.config.json
Comment (ghost) [reply packages/contracts/slither.config.json]: all though on adding it was still giving the error as before but it was in blue so can be taken as suggestion

File: packages/contracts/src/dollar/utils/CollectableDust.sol
Comment (ghost) [reply packages/contracts/src/dollar/utils/CollectableDust.sol:1]: but it was a blue error couldn't cause any problem we can get that file back if we its needed

File: packages/contracts/slither.config.json
Comment (0x4007) [reply packages/contracts/slither.config.json]: Perhaps we can initialize these values with `0` address and throw an error if they are the `0` address. That way during runtime we should never have an issue, and Slither might stop complaining?

File: packages/contracts/src/dollar/utils/CollectableDust.sol
Comment (0x4007) [reply packages/contracts/src/dollar/utils/CollectableDust.sol:1]: Collectable dust is the only way we're able to recover funds in our smart contracts in case some issues in the code make funds "unrecoverable"

Comment (ghost) [reply packages/contracts/src/dollar/utils/CollectableDust.sol:1]: > Collectable dust is the only way we're able to recover funds in our smart contracts in case some issues in the code make funds "unrecoverable"

ohh im adding it back btw imma commit it soon

File: packages/contracts/slither.config.json
Comment (ghost) [reply packages/contracts/slither.config.json]: > Perhaps we can initialize these values with `0` address and throw an error if they are the `0` address. That way during runtime we should never have an issue, and Slither might stop complaining?

oh sure we can try

**Comment by ghost** [comment]
@pavlovcik i added the collectable dust files back it seems to passed the tests too

**Comment by 0x4007** [comment]
I added comments where I think I can add value. The Slither stuff is a bit out of my wheelhouse, perhaps I can get another approval in here.

File: packages/contracts/slither.config.json
Comment (ghost) [reply packages/contracts/slither.config.json]: > Perhaps we can initialize these values with `0` address and throw an error if they are the `0` address. That way during runtime we should never have an issue, and Slither might stop complaining?

but like if we initialize the variable which is constant to 0 how we gonna change it 🤔

Comment (ghost) [reply packages/contracts/slither.config.json]: @pavlovcik i used immutable and it seemss to worked should we go with it?

Comment (ghost) [reply packages/contracts/slither.config.json]: slither not complaining too and also removed the "state-variables-that-could-be-declared-constant"

Comment (0x4007) [reply packages/contracts/slither.config.json]: I'm not a Solidity developer so I don't know.

Comment (ghost) [reply packages/contracts/slither.config.json]: > I'm not a Solidity developer so I don't know.

oh using the immutable then constant is like immutable's can be changed only using Constructors making it kinda constant but  slither seems to have to no problem with it

Comment (ghost) [reply packages/contracts/slither.config.json]: > @pavlovcik i used immutable and it seemss to worked should we go with it?

@rndquu

**Comment by ghost** [comment]
@rndquu can u take a look pls

**Review by 0x4007 — APPROVED** [review]
It looks good to me but I'm not a Solidity developer.

File: packages/contracts/slither.config.json
Comment (rndquu) [reply packages/contracts/slither.config.json]: @AnakinSkywalkeer I can't find the `state-variables-that-could-be-declared-constant` detector docs [here](https://github.com/crytic/slither/wiki/Detector-Documentation). There is [one](https://github.com/crytic/slither/wiki/Detector-Documentation#state-variables-that-could-be-declared-constant) called `constable-states` which seems to do the same task. Could you share a URL for the `state-variables-that-could-be-declared-constant` detector docs?

File: packages/contracts/src/dollar/core/ERC1155Ubiquity.sol
Comment (rndquu) [inline packages/contracts/src/dollar/core/ERC1155Ubiquity.sol]: Why was this method added? It is `public` and **adds balance to any address**.

File: packages/contracts/src/dollar/core/ERC1155Ubiquity.sol
Comment (rndquu) [inline packages/contracts/src/dollar/core/ERC1155Ubiquity.sol]: Why do we need these methods: `accessTotalSupply()`, `_incrementTotalSupply()`, `_decrementTotalSupply()`?

File: packages/contracts/src/dollar/core/ERC1155Ubiquity.sol
Comment (rndquu) [inline packages/contracts/src/dollar/core/ERC1155Ubiquity.sol]: Can't we make it public to eliminate [such refactors](https://github.com/ubiquity/ubiquity-dollar/pull/681/files#diff-cae9a99180379f71b4ba2e7d815f51849fcdbe1b33748736aeb243cd046c491fR102)?

**Review by rndquu — CHANGES_REQUESTED** [review]
1. The slither CI workflow still [fails](https://github.com/rndquu/ubiquity-dollar/actions/runs/5309480450/jobs/9610186891), pls fix
2. Pls enable the [slither CI workflow](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/workflows/slither-analysis.yml) to be executed on `push` and `pull_request`

File: packages/contracts/src/dollar/core/ERC1155Ubiquity.sol
Comment (ghost) [reply packages/contracts/src/dollar/core/ERC1155Ubiquity.sol]: > Can't we make it public to eliminate [such refactors](https://github.com/ubiquity/ubiquity-dollar/pull/681/files#diff-cae9a99180379f71b4ba2e7d815f51849fcdbe1b33748736aeb243cd046c491fR102)?

for sure we can make it public should i go for it?

File: packages/contracts/src/dollar/utils/CollectableDust.sol
Comment (rndquu) [reply packages/contracts/src/dollar/utils/CollectableDust.sol:1]: We moved to the diamond storage so now we're using the [CollectableDustFacet](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/facets/CollectableDustFacet.sol) contract instead of [CollectableDust](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/utils/CollectableDust.sol).

Removing [CollectableDust](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/utils/CollectableDust.sol) is not part of this issue so it's ok to keep it in this PR.

File: packages/contracts/slither.config.json
Comment (ghost) [reply packages/contracts/slither.config.json]: > @AnakinSkywalkeer I can't find the `state-variables-that-could-be-declared-constant` detector docs [here](https://github.com/crytic/slither/wiki/Detector-Documentation). There is [one](https://github.com/crytic/slither/wiki/Detector-Documentation#state-variables-that-could-be-declared-constant) called `constable-states` which seems to do the same task. Could you share a URL for the `state-variables-that-could-be-declared-constant` detector docs?

for sure , [here](https://github.com/crytic/slither/wiki/Detector-Documentation#state-variables-that-could-be-declared-constant)

File: packages/contracts/src/dollar/core/ERC1155Ubiquity.sol
Comment (rndquu) [reply packages/contracts/src/dollar/core/ERC1155Ubiquity.sol]: I think it's ok to make it `public` or `internal` and to remove these methods: `accessTotalSupply(), _incrementTotalSupply(), _decrementTotalSupply()`

File: packages/contracts/slither.config.json
Comment (rndquu) [reply packages/contracts/slither.config.json]: The detector you mentioned in the URL is called `constable-states` while in the code the `state-variables-that-could-be-declared-constant` detector is used which is basically an invalid detector name. Perhaps we can remove the `state-variables-that-could-be-declared-constant` detector from the exclude list?

Comment (ghost) [reply packages/contracts/slither.config.json]: > The detector you mentioned in the URL is called `constable-states` while in the code the `state-variables-that-could-be-declared-constant` detector is used which is basically an invalid detector name. Perhaps we can remove the `state-variables-that-could-be-declared-constant` detector from the exclude list?

yeah sure

File: packages/contracts/src/dollar/core/ERC1155Ubiquity.sol
Comment (ghost) [reply packages/contracts/src/dollar/core/ERC1155Ubiquity.sol]: > I think it's ok to make it `public` or `internal` and to remove these methods: `accessTotalSupply(), _incrementTotalSupply(), _decrementTotalSupply()`

sure yeah  that methods can be removed then imma do it

**Comment by ghost** [comment]
@rndquu enabled the slither analysis

File: packages/contracts/src/dollar/utils/CollectableDust.sol
Comment (AnakinSkywalkeer) [reply packages/contracts/src/dollar/utils/CollectableDust.sol:1]: > We moved to the diamond storage so now we're using the [CollectableDustFacet](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/facets/CollectableDustFacet.sol) contract instead of [CollectableDust](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/utils/CollectableDust.sol).
> 
> Removing [CollectableDust](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/utils/CollectableDust.sol) is not part of this issue so it's ok to keep it in this PR.

oh ugh i added it back that means imma just remove them again?

**Comment by ghost** [comment]
Ah what making the slither workflow to throw error like is these errors gonna make it throw error? isn't these should be ignored or the warnings above the INFO:Detectors: are making it to break hmm
image

**Comment by rndquu** [comment]
> Ah what making the slither workflow to throw error like is these errors gonna make it throw error? isn't these should be ignored or the warnings above the INFO:Detectors: are making it to break hmm image

I guess this bunch of errors makes the build to fail:
```
Pragma version0.8.19 (src/dollar/utils/SafeAddArray.sol#3) necessitates a version too recent to be trusted. Consider deploying with 0.8.18.
CollectableDustFacet (src/dollar/facets/CollectableDustFacet.sol#7-24) should inherit from ICollectableDust (src/dollar/interfaces/utils/ICollectableDust.sol#4-14)
Constants.admin (scripts/deploy/dollar/solidityScripting/00_Constants.s.sol#13) should be constant
...
```

**Comment by ghost** [comment]
> > Ah what making the slither workflow to throw error like is these errors gonna make it throw error? isn't these should be ignored or the warnings above the INFO:Detectors: are making it to break hmm image
> 
> I guess this bunch of errors makes the build to fail:
> 
> ```
> Pragma version0.8.19 (src/dollar/utils/SafeAddArray.sol#3) necessitates a version too recent to be trusted. Consider deploying with 0.8.18.
> CollectableDustFacet (src/dollar/facets/CollectableDustFacet.sol#7-24) should inherit from ICollectableDust (src/dollar/interfaces/utils/ICollectableDust.sol#4-14)
> Constants.admin (scripts/deploy/dollar/solidityScripting/00_Constants.s.sol#13) should be constant
> ...
> ```

um like i got "incorrect-versions-of-solidity" in the slither.json to be excluded it should not create problem 🤔

**Comment by ghost** [comment]
hm lets try removing these errors too ig may be it gonna fix the issue btw can we use immutable instead of constant cuz like 
  uint256 deployerPrivateKey = vm.envUint("PRIVATE_KEY");
    address constant admin = vm.envAddress("PUBLIC_KEY");
    can't really be the constant cuz the values are not known at the compile time cuz of that im asking 
```INFO:Detectors:
Constants.admin (scripts/deploy/dollar/solidityScripting/00_Constants.s.sol#13) should be constant 
Constants.deployerPrivateKey (scripts/deploy/dollar/solidityScripting/00_Constants.s.sol#12) should be constant 
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#state-variables-that-could-be-declared-constant```

**Comment by rndquu** [comment]
> hm lets try removing these errors too ig may be it gonna fix the issue btw can we use immutable instead of constant cuz like uint256 deployerPrivateKey = vm.envUint("PRIVATE_KEY"); address constant admin = vm.envAddress("PUBLIC_KEY"); can't really be the constant cuz the values are not known at the compile time cuz of that im asking
> 
> ```
> Constants.admin (scripts/deploy/dollar/solidityScripting/00_Constants.s.sol#13) should be constant 
> Constants.deployerPrivateKey (scripts/deploy/dollar/solidityScripting/00_Constants.s.sol#12) should be constant 
> Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#state-variables-that-could-be-declared-constant```
> ```

you can use whatever fixes the error

**Comment by ghost** [comment]
@rndquu yea so i was seems to be using wrong detector for avoiding the slither errors about solidity version now it seems to be workin and passed the checks

File: packages/contracts/src/dollar/core/ERC1155Ubiquity.sol
Comment (rndquu) [inline packages/contracts/src/dollar/core/ERC1155Ubiquity.sol]: Usually variables with an `_` prefix are meant to be private. Here we convert visibility from `private` to `public` so we don't need the `_` prefix anymore.

File: packages/contracts/src/dollar/facets/CollectableDustFacet.sol
Comment (rndquu) [inline packages/contracts/src/dollar/facets/CollectableDustFacet.sol]: Why was the `ICollectableDust` interface removed? I think it is crucial that `CollectableDustFacet` should implement `ICollectableDust`.

**Review by rndquu — CHANGES_REQUESTED** [review]

File: packages/contracts/src/dollar/core/ERC1155Ubiquity.sol
Comment (ghost) [reply packages/contracts/src/dollar/core/ERC1155Ubiquity.sol]: > Usually variables with an `_` prefix are meant to be private. Here we convert visibility from `private` to `public` so we don't need the `_` prefix anymore.

sure

File: packages/contracts/src/dollar/facets/CollectableDustFacet.sol
Comment (ghost) [reply packages/contracts/src/dollar/facets/CollectableDustFacet.sol]: > Why was the `ICollectableDust` interface removed? I think it is crucial that `CollectableDustFacet` should implement `ICollectableDust`.

ohh imma add it then

File: packages/contracts/src/dollar/core/ERC1155Ubiquity.sol
Comment (ghost) [reply packages/contracts/src/dollar/core/ERC1155Ubiquity.sol]: > Usually variables with an `_` prefix are meant to be private. Here we convert visibility from `private` to `public` so we don't need the `_` prefix anymore.

there is a function called totalSupply which is causing problem cuz the variable above has the same name

Comment (ghost) [reply packages/contracts/src/dollar/core/ERC1155Ubiquity.sol]:
```
function totalSupply() public view virtual returns (uint256) {
        return totalSupply;
    }

Comment (rndquu) [reply packages/contracts/src/dollar/core/ERC1155Ubiquity.sol]: perhaps we could keep the `_` and transform `public` to `internal` or simply remove the `totalSupply()` method because public state variables have a default getter

Comment (ghost) [reply packages/contracts/src/dollar/core/ERC1155Ubiquity.sol]: > perhaps we could keep the `_` and transform `public` to `internal` or simply remove the `totalSupply()` method because public state variables have a default getter

oh sure lets remove the totalSupply() function it seems more reasonable

**Comment by ghost** [comment]
@rndquu slither is not running can u take a look?

**Comment by ghost** [comment]
okay its working now seems like it was problem on the github side

**Review by rndquu — APPROVED** [review]

**Comment by rndquu** [comment]
@zgorizzo69 pls check this PR

**Comment by rndquu** [comment]
@AnakinSkywalkeer could you resolve conflicts?

**Comment by ghost** [comment]
> @AnakinSkywalkeer could you resolve conflicts?

there was seems to be due to yarn.lock i did the change lets see if it gets fixed

**Comment by ghost** [comment]
its failing let me fix

**Comment by ghost** [comment]
@rndquu i fixed it

**Review by rndquu — APPROVED** [review]

**Comment by 0x4007** [comment]
I still see warnings. Also shouldn't it be installing Solidity `0.8.19`? 

https://github.com/ubiquity/ubiquity-dollar/actions/runs/5343061138/jobs/9685652437

I just reviewed the bounty so we can handle warnings another time.

**Comment by ghost** [comment]
> I still see warnings. Also shouldn't it be installing Solidity `0.8.19`?
> 
> https://github.com/ubiquity/ubiquity-dollar/actions/runs/5343061138/jobs/9685652437
> 
> I just reviewed the bounty so we can handle warnings another time.

oh yes the solc version is set to 0.8.16 hm should i change it?

image

**Comment by 0x4007** [comment]
I think it makes sense to match our smart contracts but I will delegate the decision to my colleagues.

**Comment by rndquu** [comment]
We don't really need `solc-select` because foundry is able to download the required solc compiler version itself and slither simply doesn't need one because it is a **static** analyzer

@AnakinSkywalkeer Could you remove these steps ([one](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/workflows/slither-analysis.yml#L22) and [two](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/workflows/slither-analysis.yml#L25)) and update the PR?

**Comment by ghost** [comment]
> We don't really need `solc-select` because foundry is able to download the required solc compiler version itself and slither simply doesn't need one because it is a **static** analyzer
> 
> 
> 
> @AnakinSkywalkeer Could you remove these steps ([one](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/workflows/slither-analysis.yml#L22) and [two](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/workflows/slither-analysis.yml#L25)) and update the PR?

Sure

**Comment by ghost** [comment]
@rndquu did it

**Comment by 0x4007** [comment]
Unless we need zgo input, please feel free to merge @rndquu

**Comment by rndquu** [comment]
> Unless we need zgo input, please feel free to merge @rndquu

let's wait for @zgorizzo69 's review

**Comment by zgorizzo69** [comment]
lgtm ! good job

**Comment by 0x4007** [comment]
Oddly enough CI fails on `development` now, even though the latest commit passed.

https://github.com/ubiquity/ubiquity-dollar/actions/runs/5382075058/jobs/9766993102

**Comment by ghost** [comment]
> Oddly enough CI fails on `development` now, even though the latest commit passed.
> 
> https://github.com/ubiquity/ubiquity-dollar/actions/runs/5382075058/jobs/9766993102

thats kinda weird

**Comment by ghost** [comment]
Running slither tests locally are passing without any errors 🤔 
image

**Pull request merged**
